packages/core/lib/testing/testsource.js

```diff
@@ -4,104 +4,104 @@ const fse = require("fs-extra");
-  const self = this;
+  async resolve(importPath) {
+    const self = this;
 
-  if (importPath === "truffle/DeployedAddresses.sol") {
-    const sourceFiles = await find_contracts(this.config.contracts_directory);
+    if (importPath === "truffle/DeployedAddresses.sol") {
+      const sourceFiles = await find_contracts(this.config.contracts_directory);
 
-    let abstractionFiles;
-    const buildDirFiles = (abstractionFiles = fse.readdirSync(
-      self.config.contracts_build_directory
-    ));
-    abstractionFiles = buildDirFiles.filter(file => file.match(/^.*.json$/));
+      let abstractionFiles;
+      const buildDirFiles = (abstractionFiles = fse.readdirSync(
+        self.config.contracts_build_directory
+      ));
+      abstractionFiles = buildDirFiles.filter(file => file.match(/^.*.json$/));
 
-    const mapping = {};
+      const mapping = {};
 
-    const blacklist = new Set(["Assert", "DeployedAddresses"]);
+      const blacklist = new Set(["Assert", "DeployedAddresses"]);
 
-    // Ensure we have a mapping for source files and abstraction files
-    // to prevent any compile errors in tests.
-    sourceFiles.forEach(file => {
-      const name = path.basename(file, ".sol");
-      if (blacklist.has(name)) return;
-      mapping[name] = false;
-    });
+      // Ensure we have a mapping for source files and abstraction files
+      // to prevent any compile errors in tests.
+      sourceFiles.forEach(file => {
+        const name = path.basename(file, ".sol");
+        if (blacklist.has(name)) return;
+        mapping[name] = false;
+      });
 
-    abstractionFiles.forEach(file => {
-      const name = path.basename(file, ".json");
-      if (blacklist.has(name)) return;
-      mapping[name] = false;
-    });
+      abstractionFiles.forEach(file => {
+        const name = path.basename(file, ".json");
+        if (blacklist.has(name)) return;
+        mapping[name] = false;
+      });
 
-    const filesData = abstractionFiles.map(file => {
-      return fse.readFileSync(
-        path.join(self.config.contracts_build_directory, file),
-        "utf8"
-      );
-    });
-
-    const addresses = filesData
-      .map(data => JSON.parse(data))
-      .map(json => contract(json))
-      .map(c => {
-        c.setNetwork(self.config.network_id);
-        if (c.isDeployed()) return c.address;
-        return null;
+      const filesData = abstractionFiles.map(file => {
+        return fse.readFileSync(
+          path.join(self.config.contracts_build_directory, file),
+          "utf8"
+        );
       });
 
-    addresses.forEach((address, i) => {
-      const name = path.basename(abstractionFiles[i], ".json");
+      const addresses = filesData
+        .map(data => JSON.parse(data))
+        .map(json => contract(json))
+        .map(c => {
+          c.setNetwork(self.config.network_id);
+          if (c.isDeployed()) return c.address;
+          return null;
+        });
```

Comment:
On the one hand this is nice, clean and easy to read. On the other hand this is :vomiting_face: since we map over the same data set 3 times..  I know this isn't new code but I wonder if it doesn't deserve a little TLC.

Reply:
Want to PR what you're thinking?